File: autocycler/kmer_graph.py
#!/usr/bin/env python3
"""
Copyright 2023 Ryan Wick (rrwick@gmail.com)
https://github.com/rrwick/Autocycler

This program is free software: you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not,
see <https://www.gnu.org/licenses/>.
"""

from .misc import reverse_complement, reverse_complement_position, iterate_fasta


class Position(object):
    """
    Position objects store the sequence, strand and position for contigs in the input assemblies.
    They are used:
    * In KmerGraph objects, where each Kmer object has a single Position.
    * In UnitigGraph objects, where each Unitig has one or more starting and ending Position
      objects on both strands. These form a doubly linked list, tracing the input contig through
      the UnitigGraph.
    """
    def __init__(self, seq_id, strand, pos):
        self.seq_id = seq_id
        self.strand = strand  # 1 for forward strand, -1 for reverse strand
        self.pos = pos  # 0-based indexing

        # Pointers to the preceding and following Position objects:
        self.prev_kmer_position = None
        self.next_kmer_position = None

    def __repr__(self):
        return f'{self.seq_id}{"+" if self.strand == 1 else "-"}{self.pos}'


class Kmer(object):
    def __init__(self, seq):
        self.seq = seq
        self.positions = []

    def __repr__(self):
        positions = ','.join([str(p) for p in sorted(self.positions)])
        return f'{positions}'

    def add_position(self, seq_id, strand, pos):
        self.positions.append(Position(seq_id, strand, pos))

    def count(self):
        return len(self.positions)


class KmerGraph(object):
    """
    This class stores all k-mers in the sequences that were added to it (a De Bruijn graph), and
    for each k-mer it stores the positions in the input sequences where that k-mer occurred.
    """
    def __init__(self, k_size):
        self.k_size = k_size
        self.kmers = {}

    def add_sequence(self, seq, seq_id):
        half_k = self.k_size // 2  # actually k_size/2 - 0.5, because k_size is odd
        for forward_pos in range(len(seq) - self.k_size + 1):
            reverse_pos = len(seq) - forward_pos - self.k_size

            forward_seq = seq[forward_pos:forward_pos+self.k_size]
            reverse_seq = reverse_complement(forward_seq)

            if forward_seq not in self.kmers:
                self.kmers[forward_seq] = Kmer(forward_seq)
                self.kmers[reverse_seq] = Kmer(reverse_seq)

            forward_centre_pos = forward_pos + half_k
            reverse_centre_pos = reverse_pos + half_k
            assert reverse_centre_pos == reverse_complement_position(forward_centre_pos, len(seq))

            self.kmers[forward_seq].add_position(seq_id, 1, forward_centre_pos)
            self.kmers[reverse_seq].add_position(seq_id, -1, reverse_centre_pos)

    def add_assemblies(self, assemblies):
        id_to_contig_info = {}
        seq_id = 0
        for assembly in assemblies:
            print(f'\nAdding {assembly} to graph:')
            for name, info, seq in iterate_fasta(assembly):
                seq_id += 1
                print(f'  {seq_id}: {name} ({len(seq)} bp)...', flush=True, end='')
                self.add_sequence(seq, seq_id)
                contig_header = name + ' ' + info if info else name
                id_to_contig_info[seq_id] = (assembly, contig_header, len(seq))
                print(' done')
        print(f'\nGraph contains {len(self.kmers)} k-mers')
        return id_to_contig_info

    def next_kmers(self, kmer):
        """
        Returns a list of the k-mers in the graph which follow the given one. The list will have a
        length from 0 to 4.
        """
        a = kmer.seq[1:] + 'A'
        c = kmer.seq[1:] + 'C'
        g = kmer.seq[1:] + 'G'
        t = kmer.seq[1:] + 'T'
        next_list = []
        if a in self.kmers:
            next_list.append(self.kmers[a])
        if c in self.kmers:
            next_list.append(self.kmers[c])
        if g in self.kmers:
            next_list.append(self.kmers[g])
        if t in self.kmers:
            next_list.append(self.kmers[t])
        assert 0 <= len(next_list) <= 4
        return next_list

    def prev_kmers(self, kmer):
        """
        Returns a list of the k-mers in the graph which precede the given one. The list will have a
        length from 0 to 4.
        """
        a = 'A' + kmer.seq[:-1]
        c = 'C' + kmer.seq[:-1]
        g = 'G' + kmer.seq[:-1]
        t = 'T' + kmer.seq[:-1]
        prev_list = []
        if a in self.kmers:
            prev_list.append(self.kmers[a])
        if c in self.kmers:
            prev_list.append(self.kmers[c])
        if g in self.kmers:
            prev_list.append(self.kmers[g])
        if t in self.kmers:
            prev_list.append(self.kmers[t])
        assert 0 <= len(prev_list) <= 4
        return prev_list

    def iterate_kmers(self):
        """
        This generator yields all the k-mers in the graph in lexographical order.
        """
        for kmer in sorted(self.kmers):
            yield self.kmers[kmer]

    def reverse(self, kmer):
        """
        Takes a Kmer object and returns its reverse complement Kmer object.
        """
        return self.kmers[reverse_complement(kmer.seq)]

File: Autocycler-Rust/src/kmer_graph.rs
// This file defines structs for building a k-mer De Bruijn graph from the input assemblies.

// Copyright 2023 Ryan Wick (rrwick@gmail.com)
// https://github.com/rrwick/Autocycler

// This file is part of Autocycler. Autocycler is free software: you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version. Autocycler
// is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
// Public License for more details. You should have received a copy of the GNU General Public
// License along with Autocycler. If not, see <http://www.gnu.org/licenses/>.

use fxhash::FxHashMap;  // a bit faster than Rust's built-in HashMap
use std::collections::hash_map::Entry;
use std::fmt;
use std::slice::from_raw_parts;

use crate::position::KmerPos;
use crate::sequence::Sequence;


pub struct Kmer {
    // Instead of storing a slice of the sequence, Kmer objects store a raw pointer to sequence.
    // This requires unsafe code to access the k-mer sequence, but it avoids a bunch of tricky
    // lifetimes.
    pointer: *const u8,
    length: usize,
    pub positions: Vec<KmerPos>,
}

impl Kmer {
    pub fn new(pointer: *const u8, length: usize, assembly_count: usize) -> Kmer {
        Kmer {
            pointer,
            length,
            positions: Vec::with_capacity(assembly_count), // most k-mers occur once per assembly
        }
    }

    pub fn seq(&self) -> &[u8] {
        let kseq = unsafe{ from_raw_parts(self.pointer, self.length) };
        kseq
    }

    pub fn add_position(&mut self, seq_id: u16, strand: bool, pos: usize) {
        let position = KmerPos::new(seq_id, strand, pos);
        self.positions.push(position);
    }

    pub fn count(&self) -> usize {
        self.positions.len()
    }

    pub fn first_position(&self, half_k: usize) -> bool {
        self.positions.iter().any(|p| p.pos as usize == half_k)
    }
}

impl fmt::Display for Kmer {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let seq = std::str::from_utf8(self.seq()).unwrap();
        let positions = self.positions.iter()
                                      .map(|p| p.to_string())
                                      .collect::<Vec<String>>()
                                      .join(",");
        write!(f, "{}:{}", seq, positions)
    }
}


pub struct KmerGraph<'a> {
    pub k_size: u32,
    pub kmers: FxHashMap<&'a [u8], Kmer>,
}

impl<'a> KmerGraph<'a> {
    pub fn new(k_size: u32) -> KmerGraph<'a> {
        KmerGraph {
            k_size,
            kmers: FxHashMap::default(),
        }
    }

    pub fn add_sequences(&mut self, seqs: &'a Vec<Sequence>, assembly_count: usize) {
        for seq in seqs {
            self.add_sequence(seq, assembly_count)
        }
    }

    pub fn add_sequence(&mut self, seq: &'a Sequence, assembly_count: usize) {
        let k_size = self.k_size as usize;
        let half_k = (self.k_size / 2) as usize;

        let forward_raw = seq.forward_seq.as_ptr();
        let reverse_raw = seq.reverse_seq.as_ptr();

        for forward_start in 0..seq.length - k_size + 1 {
            let reverse_start = seq.length - forward_start - k_size;
            let forward_end = forward_start + k_size;
            let reverse_end = reverse_start + k_size;

            let forward_k = &seq.forward_seq[forward_start..forward_end];
            let reverse_k = &seq.reverse_seq[reverse_start..reverse_end];

            match self.kmers.entry(forward_k) {
                Entry::Occupied(mut entry) => {
                    entry.get_mut().add_position(seq.id, true, forward_start + half_k);
                },
                Entry::Vacant(entry) => {
                    let mut kmer = unsafe { Kmer::new(forward_raw.add(forward_start), k_size, assembly_count) };
                    kmer.add_position(seq.id, true, forward_start + half_k);
                    entry.insert(kmer);
                }
            }

            match self.kmers.entry(reverse_k) {
                Entry::Occupied(mut entry) => {
                    entry.get_mut().add_position(seq.id, false, reverse_start + half_k);
                },
                Entry::Vacant(entry) => {
                    let mut kmer = unsafe { Kmer::new(reverse_raw.add(reverse_start), k_size, assembly_count) };
                    kmer.add_position(seq.id, false, reverse_start + half_k);
                    entry.insert(kmer);
                }
            }
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kmer() {
        let seq = String::from("ACGACTGACATCAGCACTGA").into_bytes();
        let raw = seq.as_ptr();
        let mut k = Kmer::new(raw, 4, 2);
        k.add_position(1, true, 123);
        k.add_position(2, false, 456);
        assert_eq!(format!("{}", k), "ACGA:1+123,2-456");
    }

    #[test]
    fn test_kmer_graph() {
        let mut kmer_graph = KmerGraph::new(4);
        let seq = Sequence::new(1, "ACGACTGACATCAGCACTGA".to_string(),
                                "assembly.fasta".to_string(), "contig_1".to_string(), 20);
        kmer_graph.add_sequence(&seq, 1);
        assert_eq!(kmer_graph.kmers.len(), 28);
    }
}
